### DISCUSSION TITLE
Can this proxy be used to route the request to available server

### DISCUSSION THREAD

**Discussion by ajayinnovative2** [abhinavsingh/proxy.py#1080] [Q&A] [answered]
Can this proxy be used, when I have multiple tcp clients trying to send request via proxy to multiple servers of same type associated with this proxy. Is it capable to route the request to the next available server. And in response path, can this proxy identify to whom I should give the response back?

**Comment by abhinavsingh** [comment]
Sorry, but it's unclear what exactly is that you are after.  Can you expand on it?   Best way to find out is to try the plugins and see what works for you https://github.com/abhinavsingh/proxy.py/tree/develop/proxy/plugin.

**Reply by abhinavsingh** [reply]
Hi @ajayinnovative2 , I am re-reading your comment today and may be I can get a sense what you are after.   Here is what I understand:

1. Multiple Clients <-> proxy.py <-> Multiple Upstreams
2. Part that's unclear to me is, whether the clients are pure TCP clients or are they HTTP clients?
3. If they are HTTP clients, you can easily achieve your task with proxy.py defaults.   You'll have to likely write a custom plugin which can manage your app state e.g. client IP to upstream proxy server IP mapping.
4. If they are TCP clients, you can still achieve your task with proxy.py core.   I can guide you through if you are interested.  It will simply involve write a custom core plugin.

I am assuming your clients are HTTP clients and hence can use proxy settings.  On that assumption marking this as the answer.  Re-open if my assumptions are wrong.

**Comment by ajayinnovative2** [comment]
They are tcp clients, such as telnet. I have written my own proxy (load balancer) which does exactly what I wanted initially. 
My proxy is capable enough to figure out the available server associated and route the request/response accordingly. 
My only concern it becomes platform specific, as I am creating thread for each tcp session on request basis. And when it reaches to maximum thread allowed, new request will be rejected. 

Still I am thinking to extend it to support millions of connections with seamless user experience.

**Comment by abhinavsingh** [comment, marked as answer]
That's awesome.  Indeed scaling will be another aspect for any codebase.  With `proxy.py` you can achieve it automagically.  Before you dig into it, here is a recap to understand whats going on:

1) Internally `proxy.py` is a TCP/UDP server.  This core component is responsible for listening on one or several port _(or sock files)_.  It doesn't know about what functionality will run on-top of a listening server.  It's job is to accept connections, delegate them to a worker class for handling the connection.

2) By default, `--work-klass` flag points to `proxy.http.HttpProtocolHandler` class.  This class knows how to handle an incoming HTTP client connection _(server or proxy is inferred at runtime and handled appropriately)_

3) For your implementation, you need a custom `--work-klass`.  Which is capable of handling and tunneling packets at TCP level.  Luckily you have a baked in starter template under `examples` directory.   Here is a little background on that :)

    a) We already have a `BaseTcpTunnelHandler`.  This is exactly what you need to build upon.  See https://github.com/abhinavsingh/proxy.py/blob/ee424ace992b5aeb386f729e60aece91f9421acf/proxy/core/base/tcp_tunnel.py#L26
    b) `HttpsConnectTunnelHandler` example is actually build on top of `BaseTcpTunnelHandler`.   Do also check https://github.com/abhinavsingh/proxy.py/blob/35e9c57af7c6e312ec6f22fad2137f5500b0ff15/examples/https_connect_tunnel.py#L21 for reference usage and how to kick start your server instance

That's it :).  You can customize your implementation as necessary while using `proxy.py` core.   All the flags related to core `--num-workers`, `--num-acceptors`, `--recv-buf-size` etc still works.   You'll only loose HTTP proxy related flags.   Do `-h` or `--help` to see what flags your app instance contains.

Lastly, do also check `skeleton` app structure https://github.com/abhinavsingh/proxy.py/tree/develop/skeleton.   You likely want to manage all this within your own repo / directory structure with custom `proxy.py` based code.   You can pin or upgrade `proxy.py` via `requirements.txt` as and when necessary, while keeping your application code isolated from `proxy.py` codebase.

Let me know how it goes :)
